City-of-Helsinki/drupal-module-helfi-tpr
# UHF-2908: Fix partial migrates

### 1. Fixed `track_changes` source plugin setting

The `Row->changed()` check has been broken for quite a while already, making migrations to update entities even when there are no new changes.  This should massively speed up any subsequent migrations and make `PARTIAL_MIGRATE` functionality pretty much obsolete.

### 2. Save unit ids only for first translation

At the moment unit ids are saved three times (once for every translation).

### 3. Use aggregated service and errand service data

I made a small service that aggregates TPR service and errand service data into a big JSON dump. Currently every instance makes ~2000 requests per migration to TPR's service endpoint. This should speed up both service migrations by eliminating need to create three requests per entity (one for each language).

See: https://github.com/City-of-Helsinki/drupal-tpr-aggregator

To test this:

- `drush updb`
- `drush cr`
- Run all migrations: `drush mim tpr_unit`, `drush mim tpr_service`, `drush mim tpr_errand_service`, `drush mim tpr_service_channel`
- Re-run  all migrations and they should be really fast
